File: src/main/java/fr/avalonlab/blog/workshop/blogpost/BlogPostRepository.java
package fr.avalonlab.blog.workshop.blogpost;

import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.mongodb.DB;
import com.mongodb.MongoClient;
import io.vavr.collection.List;
import org.bson.types.ObjectId;
import org.jongo.Jongo;
import org.jongo.MongoCollection;
import org.jongo.MongoCursor;
import org.jongo.marshall.jackson.JacksonMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import static org.jongo.Oid.withOid;

@Singleton
public class BlogPostRepository {

    private final Logger log = LoggerFactory.getLogger(getClass());

    private Jongo jongo;

    @Inject
    public BlogPostRepository(@Named("dbName") String dbName, MongoClient mongoClient) {
        DB mongoDb = mongoClient.getDB(dbName);

        jongo = new Jongo(mongoDb,
                new JacksonMapper.Builder()
                        .registerModule(new Jdk8Module())
                        .registerModule(new JavaTimeModule())
                        .build());
    }

    private MongoCollection getCollection() {
        return  jongo.getCollection("blogposts");
    }

    public List<BlogPost> findAll() {
        MongoCursor<BlogPost> result = getCollection().find().as(BlogPost.class);

        return List.ofAll(result);
    }


    public BlogPost findById(String id) {
        ObjectId query = new ObjectId(id);

        return getCollection().findOne(query).as(BlogPost.class);
    }

    public BlogPost create(BlogPost entity) {
        getCollection().save(entity);

        return entity;
    }

    public BlogPost update(BlogPost entity) {

        getCollection().update(withOid(entity.get_id())).with(entity);

        return entity;
    }

    public void removeById(String id) {
        ObjectId query = new ObjectId(id);

        getCollection().remove(query);
    }

    public void removeAll(){
        getCollection().drop();
    }

    public List<BlogPost> find(String field, String param) {
        MongoCursor<BlogPost> result = getCollection().find("{" + field +": #}", param).as(BlogPost.class);

        return List.ofAll(result);
    }
}

File: src/main/kotlin/fr/avalonlab/blog/workshop/blogpost/BlogPostRepository.kt
package fr.avalonlab.blog.workshop.blogpost

import co.elastic.apm.api.CaptureSpan
import com.mongodb.ConnectionString
import com.mongodb.client.MongoDatabase
import org.litote.kmongo.*
import javax.inject.Inject
import javax.inject.Named

class BlogPostRepository @Inject constructor (@Named("db") db: String, @Named("dbName") dbName: String) {

    private val database: MongoDatabase
    private val BLOGPOSTS = "blogposts"

    init {
        val client = KMongo.createClient(ConnectionString(db))
        database = client.getDatabase(dbName)
    }

    private fun getCollection() = database.getCollection<BlogPost>(BLOGPOSTS)

    @CaptureSpan
    fun findAll(): List<BlogPost> {
        val result = getCollection().find()
        return result.toList()
    }

    @CaptureSpan
    fun findById(id: String): BlogPost? {
        return getCollection().findOneById(id)
    }

    @CaptureSpan
    fun create(entity: BlogPost): BlogPost {
        getCollection().insertOne(entity)
        return entity
    }

    fun update(entity: BlogPost): BlogPost {
        getCollection().updateOne(entity)
        return entity
    }

    fun removeById(id: String) {
        //val query = ObjectId(id)
        getCollection().deleteOneById(id)
    }

    fun removeAll() {
        getCollection().drop()
    }

    fun find(field: String, param: String): List<BlogPost> {
        val result = getCollection().find("{ $field: '$param' }")
        return result.toList()
    }
}